ros/src/system/autoware_health_checker/src/diag_buffer.cpp

```diff
@@ -0,0 +1,93 @@
+        data.status.insert(data.status.end(),buffer_[autoware_health_checker::LEVEL_WARN].status.begin(),buffer_[autoware_health_checker::LEVEL_WARN].status.end());
+        data.status.insert(data.status.end(),buffer_[autoware_health_checker::LEVEL_OK].status.begin(),buffer_[autoware_health_checker::LEVEL_OK].status.end());
+        data.status.insert(data.status.end(),buffer_[autoware_health_checker::LEVEL_UNDEFINED].status.begin(),buffer_[autoware_health_checker::LEVEL_UNDEFINED].status.end());
+        std::sort(data.status.begin(), data.status.end(), std::bind(&DiagBuffer::compareTimestamp, this, std::placeholders::_1, std::placeholders::_2));
+        buffer_.clear();
+        return data;
+    }
+
+    uint8_t DiagBuffer::getErrorLevel()
+    {
+        std::lock_guard<std::mutex> lock(mtx_);
+        updateBuffer();
+        if(buffer_[autoware_health_checker::LEVEL_FATAL].status.size() != 0)
+        {
+            return autoware_health_checker::LEVEL_FATAL;
+        }
+        else if(buffer_[autoware_health_checker::LEVEL_ERROR].status.size() != 0)
+        {
+            return autoware_health_checker::LEVEL_ERROR;
+        }
+        else if(buffer_[autoware_health_checker::LEVEL_WARN].status.size() != 0)
+        {
+            return autoware_health_checker::LEVEL_WARN;
+        }
+        else
+        {
+            return autoware_health_checker::LEVEL_OK;
+        }
+    }
+
+    // filter data from timestamp and level
+    autoware_system_msgs::DiagnosticStatusArray DiagBuffer::filterBuffer(ros::Time now, uint8_t level)
+    {
+        autoware_system_msgs::DiagnosticStatusArray filterd_data;
+        autoware_system_msgs::DiagnosticStatusArray ret;
+        if(buffer_.count(level) != 0)
+        {
+            filterd_data = buffer_[level];
+        }
+        for(auto data_itr = filterd_data.status.begin(); data_itr != filterd_data.status.end(); data_itr++)
+        {
+            if(data_itr->header.stamp> (now - buffer_length_))
+            {
+                ret.status.push_back(*data_itr);
+            }
+        }
+        return ret;
+    }
+
+    void DiagBuffer::updateBuffer()
+    {
+        ros::Time now = ros::Time::now();
+        buffer_[autoware_health_checker::LEVEL_FATAL] = filterBuffer(now, autoware_health_checker::LEVEL_FATAL);
+        buffer_[autoware_health_checker::LEVEL_ERROR] = filterBuffer(now, autoware_health_checker::LEVEL_ERROR);
+        buffer_[autoware_health_checker::LEVEL_WARN] = filterBuffer(now, autoware_health_checker::LEVEL_WARN);
+        buffer_[autoware_health_checker::LEVEL_OK] = filterBuffer(now, autoware_health_checker::LEVEL_OK);
+        buffer_[autoware_health_checker::LEVEL_UNDEFINED] = filterBuffer(now, autoware_health_checker::LEVEL_UNDEFINED);
+        return;
+    }
+
+    bool DiagBuffer::compareTimestamp(const autoware_system_msgs::DiagnosticStatus &a, const autoware_system_msgs::DiagnosticStatus &b)
```

Comment:
I think that it is not clear whether it is `<` or `>` as seen from the name.
For example, it should be `isOlderTimestamp(a, b)`.

Reply:
Renamed